servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+import io.servicetalk.buffer.netty.BufferUtils;
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Publisher;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.MessageToByteEncoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.netty.util.internal.PlatformDependent.throwException;
+import static io.servicetalk.buffer.api.EmptyBuffer.EMPTY_BUFFER;
+import static io.servicetalk.buffer.api.ReadOnlyBufferAllocators.DEFAULT_RO_ALLOCATOR;
+import static io.servicetalk.buffer.netty.BufferUtils.getByteBufAllocator;
+import static io.servicetalk.buffer.netty.BufferUtils.newBufferFrom;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static java.util.Objects.requireNonNull;
+
+final class NettyChannelContentCodec extends AbstractContentCodec {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyChannelContentCodec.class);
+    private static final Buffer END_OF_STREAM = DEFAULT_RO_ALLOCATOR.fromAscii(" ");
+
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier;
+    private final Supplier<ByteToMessageDecoder> decoderSupplier;
+
+    NettyChannelContentCodec(final CharSequence name,
+                             final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier,
+                             final Supplier<ByteToMessageDecoder> decoderSupplier) {
+        super(name);
+        this.encoderSupplier = encoderSupplier;
+        this.decoderSupplier = decoderSupplier;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        final int availableBytes = src.readableBytes() - offset;
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0");
+        }
+
+        if (length > availableBytes) {
+            throw new IllegalStateException("Invalid length: " + length + " (expected <= " + availableBytes);
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeOutbound(origin);
+
+            // May produce footer
+            preparePendingData(channel);
+            return requireNonNull(drainChannelOutToSingleBuffer(channel, allocator));
```

Comment:
The NPE here will be hard to understand. Consider using a special exception type for encoding and decoding (one or separate) to signal any problems associated with compression. We will use this exception type in server filter to return 413.

Reply:
Added two new Exceptions in the api module. We can further enhance them later as needed to have more info as to what codec was in use etc.